### PR TITLE
Simplification of application-state code

### PR DESCRIPTION
**Pull request by bstoeger**
### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

This tries to make the app-state thing a bit more explicit and less dynamic.

The backstory: It's three times that I had to figure our how the widgets are disabled in dive-site-edit mode. It's clever, but very hard to follow: dynamic property lists in arrays. And since we only use a single property (enabled) over-engineered.

I replaced that by some old-fashioned flags and Boolean expressions. Boring, yes.

Moreover, I replaced the QByteArray index by a simple enum. Thus, typos will be caught at runtime.

The use of an `enum class` construct may be questionable to some, as it means that it has to be cast to int to be used as an array index. We could also use something like:

```
struct ApplicationState {
	enum State {
		...
	};
};
```
The disadvantage here would be that the type of the enum is `ApplicationState::State`. I'm a bit undecided on that point.

Overall, the new code is not very pretty, but at least to me it is more palatable. YMMV.

### Changes made:

1) Turn application state index from QByteArray to enum.
2) Use boring flags instead of property lists to enable widgets.
3) Use function instead of macro to set the individual quadrants.

### CODE REVIEW

File: core/applicationstate.cpp [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#include "applicationstate.h"
```
Comment (dirkhh) [inline core/applicationstate.cpp:2]: we like having an SPDX comment on the top of all files

File: core/applicationstate.cpp [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+static ApplicationState appState = (ApplicationState)-1; // Set to an invalid value
```
Comment (dirkhh) [inline core/applicationstate.cpp:4]: space around the `-` ?

File: core/applicationstate.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+// to integer types if e.g. used as array-indices.
+enum class ApplicationState {
+	Default,
+	EditDive,
+	PlanDive,
+	EditPlannedDive,
+	EditDiveSite,
+	FilterDive,
+	Count
+};
```
Comment (dirkhh) [inline core/applicationstate.h:16]: I don't think we are consistent with this - don't we use all caps for things like this in other cases? For example in `enum cylinderuse`...

Looking further down I realize why you are using mixed case. This makes the changes smaller (or at least easier). I still think all caps would be better.

File: core/qthelper.cpp [hunk 4]
```diff
@@ -1362,21 +1363,9 @@ extern "C" void parse_display_units(char *line)
...
-
-void setCurrentAppState(const QByteArray &state)
-{
-	currentApplicationState = state;
-}
-
 extern "C" bool in_planner()
 {
-	return currentApplicationState == "PlanDive" || currentApplicationState == "EditPlannedDive";
+	return getAppState() == ApplicationState::PlanDive || getAppState() == ApplicationState::EditPlannedDive;
```
Comment (dirkhh) [inline core/qthelper.cpp:1368]: I like this much better.

File: desktop-widgets/mainwindow.cpp [hunk 5]
```diff
@@ -369,17 +369,17 @@ void MainWindow::setupSocialNetworkMenu()
 {
 }
 
-void MainWindow::setStateProperties(const QByteArray& state, const PropertyList& tl, const PropertyList& tr, const PropertyList& bl, const PropertyList& br)
+void MainWindow::setStateProperties(ApplicationState state, const PropertyList& tl, const PropertyList& tr, const PropertyList& bl, const PropertyList& br)
 {
-	stateProperties[state] = PropertiesForQuadrant(tl, tr, bl, br);
+	stateProperties[(int)state] = PropertiesForQuadrant(tl, tr, bl, br);
```
Comment (dirkhh) [inline desktop-widgets/mainwindow.cpp]: agreed, that's a bit awkward, but... meh. Overall it's a win.

File: desktop-widgets/mainwindow.cpp [hunk 6]
```diff
@@ -1835,9 +1822,15 @@ void MainWindow::showFilterIfEnabled()
 		ui.bottomSplitter->setSizes({ EXPANDED, COLLAPSED });
 	}
 }
-void MainWindow::registerApplicationState(ApplicationState state, QWidget *topLeft, QWidget *topRight, QWidget *bottomLeft, QWidget *bottomRight)
+void MainWindow::registerApplicationState(ApplicationState state, QWidget *topLeft, QWidget *topRight, QWidget *bottomLeft, QWidget *bottomRight,
+					  int topLeftFlags, int topRightFlags, int bottomLeftFlags, int bottomRightFlags)
```
Comment (dirkhh) [inline desktop-widgets/mainwindow.cpp]: I know this is a weird question... but what makes more sense:
`(state, widget1, widget2, widget3, widget4, flag1, flag2, flag3, flag4)`
or
`(state, widget1, flag1, widget2, flag2, widget3, flag3, widget4, flag4)`
?
For some reason I first thought the second one would be better. But after writing this, I'm not sure anymore. So consider this food for thought...

File: desktop-widgets/mainwindow.cpp [hunk 7]
```diff
@@ -1835,9 +1822,15 @@ void MainWindow::showFilterIfEnabled()
...
 {
-	applicationState[(int)state] = WidgetForQuadrant(topLeft, topRight, bottomLeft, bottomRight);
+	applicationState[(int)state] = {
+		{ topLeft, topLeftFlags},
+		{ topRight, topRightFlags},
+		{ bottomLeft, bottomLeftFlags},
+		{ bottomRight, bottomRightFlags}
```
Comment (dirkhh) [inline desktop-widgets/mainwindow.cpp]: And this is why I thought the first order was better... because we use them in these pairs...

**Review by dirkhh — CHANGES_REQUESTED** [review]
Mostly questions, but at least one actual change request (the missing SPDX header)

File: core/applicationstate.cpp (hunk 1 above)
Comment (bstoeger) [reply core/applicationstate.cpp:2]: Done.

File: core/applicationstate.cpp (hunk 2 above)
Comment (bstoeger) [reply core/applicationstate.cpp:4]: That's a cast and a unary '-'. We usually don't add space for either of those. This could also be written as:
```
static ApplicationState appState = ApplicationState(-1); // Set to an invalid value
```
Perhaps that's more clear?

File: core/applicationstate.h (hunk 3 above)
Comment (bstoeger) [reply core/applicationstate.h:16]: The reason I didn't got wall all-caps is that normal enums work more like macros - the values are valid in the outer scope. I was under the impression that that's why typically all-caps are used: to avoid name conflicts. Here it's more like member-variables.

But I can change to all-caps, no problem. Up to you.

File: desktop-widgets/mainwindow.cpp (hunk 6 above)
Comment (bstoeger) [reply desktop-widgets/mainwindow.cpp]: First I was also going with `state, widget` pairs. But then I thought these lines:
```
        registerApplicationState(ApplicationState::Default, mainTab.get(), profileContainer, diveList, mapWidget,
                                                            FLAG_NONE, FLAG_NONE, FLAG_NONE, FLAG_NONE);
```
were more in line with the old code. I'm perfectly fine with changing it.

Comment (bstoeger) [reply desktop-widgets/mainwindow.cpp]: Well, I just realized that when going the (widget, flag) way, we can arrange the quadrants in the source code like they are on the screen. Pure genius - let's do this. 😀

File: core/applicationstate.cpp (hunk 2 above)
Comment (dirkhh) [reply core/applicationstate.cpp:4]: :faceplant:
Yeah. Right.
Not having the space is the correct thing.

File: core/applicationstate.h (hunk 3 above)
Comment (dirkhh) [reply core/applicationstate.h:16]: That makes sense. I'm fine with it.

File: desktop-widgets/mainwindow.cpp (hunk 6 above)
Comment (dirkhh) [reply desktop-widgets/mainwindow.cpp]: 💯

File: desktop-widgets/mainwindow.cpp [hunk 8]
```diff
@@ -171,26 +171,18 @@ MainWindow::MainWindow() : QMainWindow(),
...
+	registerApplicationState(ApplicationState::EditDive, { { mainTab.get(), FLAG_NONE }, { profileContainer, FLAG_NONE },
+							       { diveList, FLAG_NONE },       { mapWidget, FLAG_NONE } });
+	registerApplicationState(ApplicationState::PlanDive, { { divePlannerWidget, FLAG_NONE },         { profileContainer, FLAG_NONE },
+							       { divePlannerSettingsWidget, FLAG_NONE }, { plannerDetails, FLAG_NONE } });
+	registerApplicationState(ApplicationState::EditPlannedDive, { { divePlannerWidget, FLAG_NONE }, { profileContainer, FLAG_NONE },
+								      { diveList, FLAG_NONE },          { mapWidget, FLAG_NONE } });
+	registerApplicationState(ApplicationState::EditDiveSite, { { diveSiteEdit, FLAG_NONE }, { profileContainer, FLAG_DISABLED },
+								   { diveList, FLAG_DISABLED }, { mapWidget, FLAG_NONE } });
+	registerApplicationState(ApplicationState::FilterDive, { { mainTab.get(), FLAG_NONE }, { profileContainer, FLAG_NONE },
+								 { diveList, FLAG_NONE },      { &filterWidget2, FLAG_NONE } });
```
Comment (dirkhh) [inline desktop-widgets/mainwindow.cpp:185]: That is awesome.

**Review by dirkhh — APPROVED** [review]
